### DISCUSSION TITLE
Trying to work out what triggers SensorStatus.java to build a new sensor_status string

### DISCUSSION THREAD

**Discussion by guydavies** [NightscoutFoundation/xDrip#3691] [Q&A]
Hi,

I've just acquired a Pixel Watch 2 and I've got xDrip+ running on my Pixel 7 phone and on the watch.  I also have Tasker running on the phone and pulling data from my xDrip in order to display to a different set of watch faces (freDiabetics' watch faces specifically).  The Tasker flows search in the output from the local web API for the first sgv entry that contains the sensor_status data, which it is faithfully pushing onto my watch.  The problem is that I use CamAPS FX with xDrip in companion mode, so there is no natively obtained Sensor Status information.  I tried sending a Sensor Start event to Nightscout, which correctly restarts the SAGE feature in the NS web UI, but xDrip doesn't seem to use that to trigger a recalculation of the sensor_status string.  So my watch says it's 1006 days and several hours since I last started a sensor.  I'm wondering if there's a particular piece of API voodoo I need for Nightscout (I am updating using some Automate flows for Insulin change, Catheter change and Sensor change each of which just does a post of those eventTypes into NS) that would trigger xDrip to recognise a new sensor has started?

Thanks for all the amazing work that you all put into xDrip over the years.  It is enormously appreciated.

Guy

**Comment by Navid200** [comment]
This may help.

In a case where xDrip acts is responsible for calibrations, it's very important to track when the sensor started.
As an example, if you still were using a Dexcom G5 and you were using it in non native mode, xDrip would be handling calibrations.

But, if you are using a Dexcom G7, you will be using xDrip in native mode.  This means that the entire CGM intelligence is in the G7 device.  
In that case, you can see when the session started by effectively asking G7 (Dex status page).   There is no valid time variable inside xDrip representing when the session started. There is no need for it.  For example, the function that shows when the session expires on the main screen, uses the start time from G7 to calculate it.

We don't maintain a start time in xDrip under all conditions.
It's not just that.  In fact, there could be an invalid start time in xDrip.
The developer using it needs to know if it is valid or not.

I don't believe we have a way of knowing when a session started in case of the companion app mode.

I am still learning and what I have stated here is only my take.

**Comment by guydavies** [comment]
Thanks Navid,

I understand why xDrip wouldn't store the last restart based on what you
describe.

It would be interesting if the same approach could be used for CamAPS FX as
you use for Dexcom G7. The sensor life time and remaining time are both
shown in the main "hamburger" menu. I wonder if a similar scrape could be
done. This would definitely make some of the available watch faces more
functional.

I'm afraid I have no experience with Java so helping with a suggested patch
would be challenging. I'll take a look if it appears straightforward but
would definitely appreciate a skilled programmer doing it so it's reliable.
I can raise a feature request if that would help.

Rgds

Guy

**Reply by Navid200** [reply]
> The sensor life time and remaining time are both shown in the main "hamburger" menu. 

Please explain what you mean by that.
You don't mean the xDrip hamburger menu, do you?

Can you show a screenshot showing where you can see the remaining sensor time?

**Comment by guydavies** [comment]
You're right. I'm sorry I wasn't clear. I meant in the CamAPS UI. See the
attached screenshot.

Rgds


Guy

**Reply by Navid200** [reply]
I see no screenshot.  But, it is good enough that you clarified you meant in the CamAPS app.

What sensor do you use with CamAPS?  Is it Dexcom or Libre?

**Comment by guydavies** [comment]
I use G6. That's all that's approved for use in France with CamAPS FX.

**Reply by Navid200** [reply]
Do you run CamAPS on the same device that runs xDrip?

**Comment by guydavies** [comment]
Yes. That's a requirement for the companion app mode.

Rgds

Guy

**Reply by Navid200** [reply]
Yes, of course.
If you were to run on a different device, there is a possible option that may or may not work: https://navid200.github.io/xDrip/docs/G6_slot.html
It is still behind engineering mode because we cannot guarantee that it would work for all devices.

The Companion App mode was created as a workaround if we couldn't collect directly.  The outcome is that we are limited to what is shown in the notifications by the companion app.  
I am not aware of a way for xDrip to get any information from a companion app other than readings.

**Comment by guydavies** [comment]
Ah. Right. I've used a similar technique in the past with the Bluejay watch
as a collector. It was not very robust as a collector and I ultimately used
it as just a means to see the BG without getting my phone out with xDrip in
companion mode on the phone.

I don't have a second device to use so that option isn't available anyway.

Thanks for the explanation. It's a shame that there is no native
notification that can be used by xDrip, but even if there were, you'd need
to keep some internal record of it and if you missed the notification it
would not be accurate.

Rgds

Guy

**Comment by jamorham** [comment]
Have you tried enabling the "companion bluetooth" setting? It doesn't work on all phones but might help. Turn it off again if it breaks things.

**Comment by guydavies** [comment]
Thanks. I had never tried that. I just turned it on and I'll let you know
if it helps.

Rgds Guy

**Comment by guydavies** [comment]
Hi again,

What's the expected behaviour (change) when turning this setting on?  The
companion mode seems to be working well (still) but no new info seems to
have been picked up. It's not obvious to me how turning on BT for the
companion mode would enable the correct information to be exposed to tasker
for SAGE if the data is indeed obtained as Navid described.

Thanks for your help.

Guy

**Comment by guydavies** [comment]
Ha!  I searched through the instructions for the companion mode (actually
searching for the BT description) and noticed that it said to stop/start
the sensor in companion mode wouldn't have any effect (probably ;-)). So I
tried it and bam! The info on the watch was updated. This is now all
working. One minor niggle was that all previous data from the watch face
was lost with the stop/start. Normally, I suspect that wouldn't matter with
the 2 hour warmup.

Rgds Guy
